### PR TITLE
add upload date to attachment details, closes #3

### LINKED ISSUE #3: Return attachments with time/date updated and unique identifier
> Attachments return a URL, a content disposition and an expiration date.  Two things that would be useful would be a date/time of when it was posted if available, to help keep track of which ones are new, as well as any sort of unique identifier if it exists, again to keep track of which ones have already been fetched.  I do not think it enforces file names to be unique, so I would like to avoid trying to rely on just the name.

### PR DESCRIPTION
**Pull request by fgregg**
This adds the upload date to the metadata of an attachment.

### CODE REVIEW

File: govqa/base.py [hunk 1]
```diff
@@ -37,14 +36,22 @@ def __init__(self, domain, username, password, *args, **kwargs):
...
+                "user-agent": "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Mobile Safari/537.36",
+            }
+        )
+
     def request(self, *args, **kwargs):
         response = super().request(*args, **kwargs)
 
         if "There was a problem serving the requested page" in response.text:
             response.status_code = 500
+            raise scrapelib.HTTPError(response)
```
Comment (fgregg) [inline govqa/base.py:50]: raise error in case we want to take advantage of scrapelib's retry code later.

File: govqa/base.py [hunk 2]
```diff
@@ -37,14 +36,22 @@ def __init__(self, domain, username, password, *args, **kwargs):
...
 
         elif "Page Temporarily Unavailable" in response.text:
-            response.status_code = 500
+            response.status_code = 503
```
Comment (fgregg) [inline govqa/base.py:53]: 503 is a good code for temporary error.

File: govqa/base.py [hunk 3]
```diff
@@ -73,8 +75,6 @@ def login(self):
             "//input[@name='__RequestVerificationToken']"
         )[0].value
 
-        headers["content-type"] = "application/x-www-form-urlencoded"
```
Comment (fgregg) [inline govqa/base.py:76]: don't need to set header here anymore.

File: govqa/base.py [hunk 4]
```diff
@@ -194,20 +201,27 @@ def get_request(self, request_id):
             )
 
         attachment_links = tree.xpath(
-            "//div[@id='dvAttachments']/descendant::div[@class='qac_attachment']/input[contains(@id, 'hdnAWSUrl')]"
+            "//div[@id='dvAttachments']/descendant::div[@class='qac_attachment']/input[contains(@id, 'hdnAWSUrl') or contains(@id, 'hdnAzureURL')]"
```
Comment (fgregg) [inline govqa/base.py:204]: it looks like sometimes the data is stored in azure, check for that too.

File: govqa/base.py [hunk 5]
```diff
@@ -194,20 +201,27 @@ def get_request(self, request_id):
...
-            request["attachments"].append(
-                {
-                    "url": link.attrib["value"],
-                    "content-disposition": metadata["response-content-disposition"][0],
-                    "expires": datetime.fromtimestamp(
-                        int(metadata["Expires"][0])
-                    ).isoformat(),
-                }
-            )
+            if "value" in link.attrib:
```
Comment (fgregg) [inline govqa/base.py:208]: since sometimes the data is in azure or aws we need to check.

File: govqa/base.py [hunk 6]
```diff
@@ -194,20 +201,27 @@ def get_request(self, request_id):
...
+                url = link.attrib["value"]
+                uploaded_at_str = link.xpath("../../../td[1]/text()")[0].strip()
```
Comment (fgregg) [inline govqa/base.py:210]: this is the core of the change that captures the upload date.

File: govqa/base.py [hunk 7]
```diff
@@ -194,20 +201,27 @@ def get_request(self, request_id):
...
+                metadata = parse_qs(urlparse(url).query)
+                request["attachments"].append(
+                    {
+                        "url": link.attrib["value"],
+                        "content-disposition": metadata["response-content-disposition"][
+                            0
+                        ],
+                        "expires": datetime.fromtimestamp(
+                            int(metadata["Expires"][0])
+                        ).isoformat(),
```
Comment (fgregg) [inline govqa/base.py:220]: @hancush, any reason why we are returning isoformat instead of python datetime objects?

**Review by hancush — APPROVED** [review]

**Pull request merged**